elisp/log4slime.el

```diff
@@ -92,6 +95,202 @@ argument, the parent effective log level (string)")
+    (let* ((form-args (or alt-args args))
+           (define (if  (cl-some #'consp form-args)
+                       'cl-defun
+                     'defun))
+           (suffix (symbol-name name))
+           (suffix (if (string-prefix-p "log4slime-backend-" suffix)
+                     (substring suffix 18) ; Length of prefix
+                   (error "Invalid name: %s" name))))
+      `(,define ,name ,form-args ,doc
+         (pcase log4slime-backend
+          ('slime (funcall (intern (concat "slime-" ,suffix))
+                   ,@(cl-remove '&optional args)))
+          ('sly (funcall (intern (concat "sly-" ,suffix))
+                 ,@(cl-loop for arg in form-args
+                    when (not (eq arg '&optional))
+                    collect (if (consp arg)
+                                (car arg)
+                              arg))))
+          (_ (error "Invalid log4slime backend: %s" log4slime-backend)))))))
+
+(log4slime-define log4slime-backend-connected-p ()
+  "Return true if the backend connection is open.")
+
+(log4slime-define log4slime-backend-current-package ()
+  "Return the common lisp package in the current context.
+If ‘slime-buffer-package’ or ‘sly-buffer-package’ (as
+appropriate) has a value then return that, otherwise search for
+and read an ‘in-package’ form.")
+
+
+(log4slime-define log4slime-backend-symbol-at-point ()
+  "Return the name of symbol at point, otherwise nil.")
+
+(log4slime-define log4slime-backend-sexp-at-point
+  ()
+  "Return the sexp at point, otherwise nil.
+For the Sly/Slynx backend, with non-nil INTERACTIVE, flash the
+rogion and also error in fon sexp can be found, unles ERRORP,
+with defaults to t, is passed as nil.  With non-nil STRINGP, only
+look for strings."
+  (&optional interactive stringp (errorp t)))
+
+(log4slime-define log4slime-backend-analyze-xrefs (xrefs)
+  "Find common filenames in XREFS.
+Return a list of (SINGLE-LOCATION File-ALIST).
+SINGLE-LOCATION is true if all xrefs point to the same location.
+FILE-ALIST is an alist of the form ((FILENAME . (XREF ...) ...).")
+
+(log4slime-define log4slime-backend-push-definition-stack ()
+  "Add point to the find-tag-marker-ring.")
+
+(log4slime-define log4slime-backend-xref.location (xref)
+  "Access slot \"location\" of backend XREF struct CL-X.")
+
+(log4slime-define log4slime-backend-length= (seq n)
+  "Return the results of (= (length SEQ N) from backend connection.")
+
+(log4slime-define log4slime-backend-current-connection ()
+  "Return the backend connection to use for lisp interaction.
+Return nil if there’s no connection.")
+
+(defun log4slime-backend-buffer-file-name ()
+  "Return the lisp file name associated with current buffer.
+If ‘slime-buffer-package’ or ‘sly-buffer-package’ (as
+appropriate) has a value then return that, otherwise search for
+and read an ‘in-package’ form."
+  (pcase log4slime-backend
+    ('slime (buffer-file-name))
+    ('sly (funcall (intern "sly-to-lisp-filename") (buffer-file-name)))))
```

Comment:
Yes, the two backend APIs are different enough that it is hard to get clean/y account for differences.

It is all the harder if both backends may be loaded at the same time.

If I understand your suggestion correctly, I could introduce a base backend class to dispatch on.  Follow that with a derived backend specific class for each of `log4slime` and `log4sly`.  An instance of one or the other class needs to be registered with the common code.  Then convert the thirteen or so functions that differ to `defgeneric`, adding a reference to that instance variable.

Common code in `log4emacs.el` (or `log4common.el`, or whatever) would get the bulk of the code that is currently in `log4sime.el`.  And `log4slime.el` and `log4sly.el` would provide the `defmethods` and a `make-instance` of their respective classes to the common code.

Do I understand your suggestion correctly?

Reply:
> Do I understand your suggestion correctly?

Yes. Do you think such an approach makes sense?

Reply:
It does make sense.  And it seems a cleaner solution.

I am willing to do this.  But I have questions as to naming and review preparation before I get started.

As to naming:
- common code in log4cl.el, or would you prefer another name?
- should we keep the `log4slime-` prefix across all three files, or use three prefixes?  Emacs guidelines simply suggest consistency.  On the other hand, melpazoid complains if the prefix does not follow package and filename.

As to review preparation, I still am getting my head around the github workflow.  This will likely be a large changeset.  How can I break this up to make it easier to review?

Reply:
> common code in log4cl.el, or would you prefer another name?

That seems fine as it matches the project name and the ASDF systems already use this naming scheme.

> should we keep the log4slime- prefix across all three files, or use three prefixes? Emacs guidelines simply suggest consistency. On the other hand, melpazoid complains if the prefix does not follow package and filename.

I don't have much experience with modern Emacs development. Keeping the prefix would be good for existing users, but I can see how that might violate the recommended style.

> This will likely be a large changeset. How can I break this up to make it easier to review?

Multiple self-contained commits help. For example, refactoring things to use a backend object and generic functions could be one commit, putting SLIME-specific functions into a separate file could be one commit, and adding SLY support could be one commit. Beyond that, I don't know how to make things easier to review.

Reply:
> > should we keep the log4slime- prefix across all three files, or use three prefixes? Emacs guidelines simply suggest consistency. On the other hand, melpazoid complains if the prefix does not follow package and filename.
> 
> I don't have much experience with modern Emacs development. Keeping the prefix would be good for existing users, but I can see how that might violate the recommended style.

Hmm.  Consistency seems more important than trying to be to pedantic.  I will avoid unnecessary renames.  Fix it later if it seems an issue.  I suspect it will not.

> > This will likely be a large changeset. How can I break this up to make it easier to review?
> 
> Multiple self-contained commits help. For example, refactoring things to use a backend object and generic functions could be one commit, putting SLIME-specific functions into a separate file could be one commit, and adding SLY support could be one commit. Beyond that, I don't know how to make things easier to review.

Good.  I will structure this as I would a patch stream.  It will take a bit to refactor and test.

If you would like to close this pull request, I will open a new one what I have something to review.  If you leave it open, I will update it.